### PR TITLE
Modernize test-suite in sparse tensors 

### PR DESCRIPTION
**Pull request by aocsa**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56446 Eliminate global usage of torch.set_default_dtype in test_autograd
* #56393 Eliminate global usage of torch.set_default_dtype in sparse test
* **#56392 Modernize test-suite in sparse tensors**

Fixes for gh-56371 and gh-56369

Differential Revision: [D27913212](https://our.internmc.facebook.com/intern/diff/D27913212)

Note: 
The following requests about testing listed [here](https://pearu.github.io/csr_tensor_support.html) 

#50937 (comment) - see modern COO testing
#50937 (comment) - avoid using numpy
#50937 (comment) - inefficient CSR samples

are being solved in this PR

### CODE REVIEW

File: test/test_sparse_csr.py
Comment (v0dro) [inline test/test_sparse_csr.py:190]: You probably want to test matvec for both int32 and int64 index types.

File: test/test_sparse_csr.py
Comment (v0dro) [inline test/test_sparse_csr.py:189]: Why only doubles? If you add `float` too it will use the `s_` routines in MKL and exercise that part of the code too.

File: test/test_sparse_csr.py
Comment (v0dro) [inline test/test_sparse_csr.py:205]: Might be better to include all types here such as complex since the conversion is supposed to be agnostic of data type of the values.

File: test/test_sparse_csr.py
Comment (aocsa) [reply test/test_sparse_csr.py:190]: ok

File: test/test_sparse_csr.py
Comment (aocsa) [reply test/test_sparse_csr.py:189]: got it

File: test/test_sparse_csr.py
Comment (aocsa) [reply test/test_sparse_csr.py:205]: I tested and this conversion only works with  `torch.testing.floating_types`. Complex support for sparse tensors is WIP  in PR #54153

File: torch/testing/_internal/common_utils.py
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:1052]: Add a comment for this function

I suggest changing its signature to `(self, size, nnz, *, device, dtype, index_dtype)` for readability

File: torch/testing/_internal/common_utils.py
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:1076]: Why such a tight range of values?

**Review by mruberry — APPROVED** [review]
Hey @aocsa! Overall this looks pretty good; I have a couple suggestions and a question, but nothing major

File: torch/testing/_internal/common_utils.py
Comment (aocsa) [reply torch/testing/_internal/common_utils.py:1076]: Some tests from `test_sparse` expect tight ranges because some test  with default eps/tolerance will fail.  So I've just replicated that for genSparseCSRTensor.

**Review by pearu — CHANGES_REQUESTED** [review]
Thanks @aocsa , this looks good.

I have several nits but most importantly, I suggest making `test_sparse_csr_constructor` method more thorough, in particular, ensure that all supported dtypes are covered as well as checking that the constructed CSR tensor have the expected dtype, device, and indices/values.

We may also want to extend these tests to other inputs such as lists and objects implementing array interface but atm this blocked by https://github.com/pytorch/pytorch/issues/56687 and https://github.com/pytorch/pytorch/pull/54187 .

File: test/test_sparse_csr.py
Comment (aocsa) [reply test/test_sparse_csr.py:18]: Yes all test methods requires specification of device argument

File: test/test_sparse_csr.py
Comment (aocsa) [reply test/test_sparse_csr.py:36]: Right now, it only supports: dtype={float32, float64, int16, int32, int64, int8, uint8}  so I updated to 

```
*torch.testing.get_all_dtypes(include_bool=False, include_half=False,
                                          include_bfloat16=False, include_complex=False)
```

File: test/test_sparse_csr.py
Comment (aocsa) [reply test/test_sparse_csr.py:86]: yeap, it better remote, besides adding `dtypes(get_all_dtypes)` will cover this case in test `test_sparse_csr_constructor`

File: torch/testing/_internal/common_utils.py
Comment (aocsa) [reply torch/testing/_internal/common_utils.py:1084]:
```
[------------------------------------------------ sparse.mm -------------------------------------------------]
                          |  gen_sparse_csr  |  gen_sparse_csr_from_coo  |  gen_sparse_csr_from_gen_coo_sparse
1 threads: ---------------------------------------------------------------------------------------------------
      (32, 100) | 0.1     |       17.9       |            4083.4         |                 1092.2             
      (32, 100) | 0.25    |       21.6       |            9828.9         |                 2254.4             
      (512, 100) | 0.1    |       65.5       |           62794.7         |                16265.3             
      (512, 100) | 0.25   |      142.7       |          166035.6         |                37583.0             
      (1024, 100) | 0.1   |      121.4       |          140723.3         |                34030.6             
      (1024, 100) | 0.25  |      281.1       |          372286.9         |                75763.5             
      (2048, 100) | 0.1   |      219.9       |          257356.1         |                65103.0             
      (2048, 100) | 0.25  |      521.3       |          657451.4         |               147426.5      
```
I compared three methods using this script [https://gist.github.com/aocsa/9594f45b336065e4f887006110f0df2b](url)   `gen_sparse_csr  |  gen_sparse_csr_from_coo  |  gen_sparse_csr_from_gen_coo_sparse`, the winner is `gen_sparse_csr` which generate directly CSR components.
  
Note that `gen_sparse_csr_from_gen_coo_sparse` which uses `genSparseTensor` is less flexible because it can't change the `index_dtype`, which in the case of CSR could be int32 or int64 compared with COO which by default is int64.

File: test/test_sparse_csr.py
Comment (aocsa) [reply test/test_sparse_csr.py:163]: the output is "[]" when `torch.has_mkl==True`,  is it expected @v0dro?

**Comment by aocsa** [comment]
> Thanks @aocsa , this looks good.
> 
> I have several nits but most importantly, I suggest making `test_sparse_csr_constructor` method more thorough, in particular, ensure that all supported dtypes are covered as well as checking that the constructed CSR tensor have the expected dtype, device, and indices/values.
> 
> We may also want to extend these tests to other inputs such as lists and objects implementing array interface but atm this blocked by #56687 and #54187 .

Thanks for the feedback @pearu, I addressed the comments and rebased this PR on top of the latest changes, polished a few things and added some tests, specifically related to factory tests. I think it is ready for final review and merging, let me know what you think. cc @mruberry

File: test/test_sparse_csr.py
Comment (aocsa) [reply test/test_sparse_csr.py]: I added some tests, specifically related to factory tests. I generated most of the possible combination, however the corresponding to `device='cuda'` still is missing and I will add the missing one in another PR.

File: test/test_sparse_csr.py
Comment (aocsa) [reply test/test_sparse_csr.py]: I think  the point is to guarantee tensor properties invariance during changes by comparing them with and expected string. I fixed it adding the expected output file and comparing the output to this

File: torch/testing/_internal/common_utils.py
Comment (pearu) [reply torch/testing/_internal/common_utils.py:1076]: @aocsa thanks for these benchmarks.

Interestingly, generating CSR samples is more efficient than generating COO samples. For instance, defining
```
def gen_sparse_coo(size, nnz, dtype=torch.float, index_dtype=torch.int64, device='cpu'):
    return genSparseTensor(size, 2, nnz, False, device, dtype)[0]
```
and disabling the last two columns, we have the following benchmark results:
```
                          |  gen_sparse_csr  |  gen_sparse_coo
1 threads: ---------------------------------------------------
      (32, 100) | 0.1     |       20.6       |        75.0    
      (32, 100) | 0.25    |       26.6       |       113.6    
      (512, 100) | 0.1    |       80.0       |       492.0    
      (512, 100) | 0.25   |      173.1       |      1198.1    
      (1024, 100) | 0.1   |      142.6       |       965.8    
      (1024, 100) | 0.25  |      326.9       |      2357.4    
      (2048, 100) | 0.1   |      267.5       |      1868.4    
      (2048, 100) | 0.25  |      642.3       |      5802.0    
```
that is, sampling COO is about 4-9x slower than a sampling of CSR.

This may indicate that COO sampling could be made more efficient, or perhaps not. For instance, COO sampler uses `torch.rand->mul->to(long)` to compute `indices` while CSR uses just 1D `torch.randint`.

File: test/test_sparse_csr.py
Comment (pearu) [inline test/test_sparse_csr.py:44]: Could you also add dtype/device checks:
```
self.assertEqual(dtype, sparse.dtype)
self.assertEqual(device, sparse.device)
```
to be sure that we get what we request.

File: test/test_sparse_csr.py
Comment (pearu) [reply test/test_sparse_csr.py:190]: What is the output when `torch.has_mkl==False`?

File: torch/testing/_internal/common_utils.py
Comment (pearu) [inline torch/testing/_internal/common_utils.py:1066]: This is a bug: `nnz_per_row` is 0 here and the result will be all zero `crow_indices`. I think what you meant is
```
crow_indices[1:nnz + 1] = 1
```

In fact, this reveals a bug in `torch.sparse_csr_tensor`: the invalidation of invariant `crow_indices[-1] == nnz` is not catched.

Comment (pearu) [reply torch/testing/_internal/common_utils.py:1066]: The sparse_csr_tensor bug is reported: https://github.com/pytorch/pytorch/issues/56856

**Review by pearu — CHANGES_REQUESTED** [review]
Thanks @aocsa ! LGTM, however, there is a bug in `random_sparse_csr`. Once this is fixed, I'll approve.

File: test/test_sparse_csr.py
Comment (aocsa) [reply test/test_sparse_csr.py:190]: The output is nothing, the expected output when `torch.has_mkl==True` is the following  
```
UserWarning: Pytorch is compiled with MKL LP64 and will convert crow_indices to int32.
UserWarning: Pytorch is compiled with MKL LP64 and will convert col_indices to int32.
```
for which I added two `self.assertIn`. 

Moreover, while I was testing this test I found a new issue when `size` parameter  is not provides to the sparse_csr_tensor factory method and when `index_dtype` is int32.  
https://github.com/pytorch/pytorch/blob/7b74c3c70a64eb0f377ff5e8a557cc0293239ce1/aten/src/ATen/native/sparse/SparseCsrTensor.cpp#L111
I fixed and  added this case to the test too.

File: torch/testing/_internal/common_utils.py
Comment (aocsa) [reply torch/testing/_internal/common_utils.py:1066]: Thanks for the feedback @pearu, I fixed the `random_sparse_csr` function. Related the invariant, yes there was not a check when the size parameter was not provided to the factory function, so I added the check, however I think that the check is `crow_indices[-1] <= nnz`  as it is described here: 

https://github.com/pytorch/pytorch/blob/267b554b6fd7d8dfdfb9885c663fbe48fb26b2f8/aten/src/ATen/native/sparse/SparseCsrTensor.cpp#L58

To support this I tested matmul operation with `scipy/sparse` and csr `matvec` pytorch implementation with mkl, both accept tensors with this kind of inputs. 
 
See test here:
https://gist.github.com/aocsa/ecdbfdc5e8d76c1703080076d1065aaa

Comment (pearu) [reply torch/testing/_internal/common_utils.py]: The check `crow_indices[-1] <= nnz` is wrong but let's resolve it in the context of issue #56856.

But in short, `crow_indices` must be increasing sequence and `diff(crow_indices)` gives the number of elements of all rows. That is, we have an invariant `sum(diff(crow_indices)) == nnz`. If you expand this expression, you'll get
```
crow_indices[1] - crow_indices[0] + crow_indices[2] - crow_indices[1] + ... + crow_indices[-1] - crow_indices[-2] == nnz
```
or after reduction:
```
crow_indices[-1] - crow_indices[0] == nnz
```
Since `crow_indices[0] = 0` by definition, this proves that `crow_indices[-1]` must always be equal to `nnz`.

This invariant is best tested for the case where `nnz < size[0]`. This case likely originates the heuristic condition `crow_indices[-1] <= nnz` but it is wrong because the content of `crow_indices` is left incomplete. To complete it, one must apply:
```
for n in range(k, size[0] + 1):
        crow_indices[n] = nnz
```

**Review by pearu — APPROVED** [review]
Thanks, @aocsa for the fixes.

I'll approve this PR despite the wrong `crow_indices[-1] <= nnz` check, see https://github.com/pytorch/pytorch/pull/56392#discussion_r620042454 . 
We'll need to resolve this incorrectness in the context of https://github.com/pytorch/pytorch/issues/56856 because the corresponding fix likely requires changes to the CSR construction algorithm and hence is OT for this PR.

@mruberry I think this PR is ready for merge (assuming that CI is fine with it). The PR has a high priority because other CSR-related PRs rely on this testing framework.

**Pull request closed without merging**
